### PR TITLE
Insight folders step2

### PR DESCRIPTION
**Pull request by dominikl**
Step 2 of the Insight integration of ROI folders:
- Fixed minor outstanding issues of #4450 (FolderData null checks and ROIService integration test)
- Fixed ROI propagation for ROIs in multiple folders /cc @pwalczysko 
- Take folder visibility into account to determine visibility of ROIs /cc @pwalczysko 
- Display multi-level hierarchy of ROI folders

### CODE REVIEW

File: components/insight/SRC/org/openmicroscopy/shoola/agents/measurement/util/roitable/ROINode.java [hunk 1]
```diff
@@ -216,6 +229,19 @@ public ROINode findChild(Coord3D shapeCoord)
...
+    /**
+     * Find the child node representing the given folder
+     * 
+     * @param folder
+     *            see above.
+     * @return see above.
+     */
+    public ROINode findChild(FolderData folder) {
+        if (folderMap.containsKey(folder))
+            return folderMap.get(folder);
```
Comment (mtbc) [inline components/insight/SRC/org/openmicroscopy/shoola/agents/measurement/util/roitable/ROINode.java]: Isn't this line the only of the three that you need?

Comment (dominikl) [inline components/insight/SRC/org/openmicroscopy/shoola/agents/measurement/util/roitable/ROINode.java]: That's true, thanks. Was just copy/pasting the other findChild() method, without really thinking about it.

**Comment by jburel** [comment]
For the methods used to retrieve folder added to the BrowserFacility, we should probably think of adding some parameters like userid even if it is not used now.
I will probably try to match when applicable the signature of the methods used to retrieve Project/Dataset etc

**Comment by dominikl** [comment]
Good point. I'll fix that within this PR. Also noticed I should have some integration tests for that as well.

**Comment by gusferguson** [comment]
@dominikl cc @pwalczysko 

Tested using OMERO.insight-5.3.0-m0-183-ea14612-ice35-b23-mac regions-omero.docker.openmicroscopy.org sample_user 

- Fixed ROI propagation for ROIs in multiple folders /cc @pwalczysko

Behaves as expected.

- Take folder visibility into account to determine visibility of ROIs /cc @pwalczysko

Behaves as expected.

Note the following:
- uncheck Visible for Folder: elliptical [2] - uncheck  Folder: right [2] - ROI ID 375 is not displayed (screenshot 1)
- check visible for  Folder: right [2] - ROI ID 375 becomes visible and Visible is checked for ROIs and  Folder: right [2] but remains unchecked for Folder: elliptical [2] (screenshot 2)

This is actually correct behaviour as there could be other ROIs in that folder that you want to remain visible - it just looks a little odd when there are no other invisible ROIs in folder. So no action needed.

- Display multi-level hierarchy of ROI folders

Two levels of folder hierarchy displays as expected in image ID: 4825

Good to merge from my POV - but probably a good idea for @pwalczysko to check as well.

id- 4825 p-tre_12_r3d_d3d-folders ome tiff

default

**Comment by pwalczysko** [comment]
Note that in the screenshots above (as well as in my own test with Insight) we have ``0`` in place of the ``id`` for a propagated shapes.

**Comment by pwalczysko** [comment]
Collapse of tree when ``Propagate`` is executed: Select a ROI inside a nested folder, click ``Propagate`` and draw the ROI over several z or t planes. Observe that in the moment you let the mouse button go, the hieararchy displayed in the ROI Manager collapses in all levels and folders -> this makes it hard to check the result of your action and do not really see the reason for it.

**Comment by pwalczysko** [comment]
Re: Visibility when Folder checked or unchecked - fully agree with the lowest-level folder visibility test and anyalysis by @gusferguson . What about the higher and highest folder level though ? Atm there is no action at all if you uncheck the visibility of e.g. Folder: by (id 245). All inside this folder will remain visible. I do not mean now that I have a ready solution for this one, but shouldn't we think of this ? (also, can a higher level folder contain ROIs as well as folders ? How do we adjust the behaviour then ?)

**Comment by pwalczysko** [comment]
In summary:
  - one possible bug https://github.com/openmicroscopy/openmicroscopy/pull/4458#issuecomment-181382176
  - one usability issue 
  - one conceptual question https://github.com/openmicroscopy/openmicroscopy/pull/4458#issuecomment-181389850

I would suggest lets investigate the bug, and possibility to fix the usability issue, but also happy for this  to be merged and start another Step for this.

**Comment by dominikl** [comment]
With respect to https://github.com/openmicroscopy/openmicroscopy/pull/4458#issuecomment-181382176 (id == 0): Just checked with a non RoiFolders build of Insight, same there. id == 0 until the ROI changes are saved.

**Comment by dominikl** [comment]
**Test**:
- Last commit, addresses https://github.com/openmicroscopy/openmicroscopy/pull/4458#issuecomment-181384147 ; make sure the table is not collapsed after a ROI is propagated through z/t.
- The two commits before just add more Folder specific methods and tests to the Gateway (to be congruent with methods like getProjects and getDatasets). Test: Code review and make sure Java integration tests pass.

**Comment by mtbc** [comment]
I wonder if it would be good to display something other than "0" to mean "doesn't yet have an ID assigned" (which I suspect is what the `0` signifies).

**Comment by gusferguson** [comment]
@dominikl 

Tested with OMERO.insight-5.3.0-m0-197-d42b269-ice35-b27-win  regions-omero.docker.openmicroscopy.org sample_user

Folder not collapsing after propagation of ROI.
Behaves as expected.
WRT this bug fix - good to merge.
Happy for this PR to be merged and other issues raised by @pwalczysko to be addressed in other PRs if that suits everyone best.
I am not hung up on the ID = 0 before save - don't think it will worry anyone really.

**Comment by pwalczysko** [comment]
Tested on Win7. All as expected. The only open question is the conceptual issue of visibility of high level folders https://github.com/openmicroscopy/openmicroscopy/pull/4458#issuecomment-181389850 - definitely something we need to push. Maybe merge and copy the https://github.com/openmicroscopy/openmicroscopy/pull/4458#issuecomment-181389850 into the next step ? @gusferguson

**Comment by dominikl** [comment]
Would be good if it could be merged. Then I can move on to step3. Can resolve the folder visibility issue in there.

**Comment by gusferguson** [comment]
@dominikl - I would be happy with that - we just need to decide what else would be in step 3. I am just putting final touches to all the right-click and other folder manipulation mockup - will be posted in the issue by EOB today with luck.

**Comment by sbesson** [comment]
The new Java integrations tests and all comments from @jburel on gh-4450 have been addressed here. As per @pwalczysko and @gusferguson recommendations, merging to allow future work to happen on subsequent PRs.

**Pull request merged**